lms/djangoapps/courseware/docs/decisions/0004-personalized-relative-dates.rst

```diff
@@ -0,0 +1,88 @@
+        write in absolute or relative dates.
+    b.  While there are no current plans to allow Studio authors to author
+        relative dates, by making the evenly-spacing only occur if there
+        is no existing relative dates on the course, we leave ourselves the
+        ability to add future studio authoring without changing ``edx-when``.
+    c.  Adding the relative dates on course-publish, and storing them in ``edx-when``,
+        rather than inferring them at run-time while reading from ``edx-when``
+        means that we have the course structure in-hand (from the course
+        publish), and that ``edx-when`` will remain a full listing of all dates
+        in the course. If, instead, the LMS inferred relative dates when they
+        weren't set in ``edx-when``, then new date-based functionality in the
+        future would need to either read through a specific api that would
+        inject the programmatic dates, or would have to infer its own dates.
+3.  No new fields would be added to XBlocks at this time to represent relative
+    dates. There is no studio authoring component for this feature, and the LMS
+    Xblock runtime would contextualize relative dates for a specific student
+    when supplying them to XBlock fields.
+4.  Update the documentation in ``edx-when`` to capture our current
+    understanding of
+    its responsiblities. To wit, ``edx-when`` exists to give the LMS a single place
+    to ask "What does this student have coming up, across one or many courses?", and
+    get a fast, efficient answer. To that end, the ``schedules`` django app should
+    be moved into ``edx-when``, and it will handle any future features related to
+    the ability to scale pacing in self-paced courses faster or slower. ``edx-when``
+    is thus both a performance optimization (because it is optimized for reading),
+    and the single dedicated owner of a particular user's schedule of course content.
+
+    However, it does not include information about course-level administrative dates,
+    such as the upgrade deadline, so that will need to be disentangled from the Schedule
+    model (or we'll need to expand ``edx-when`` with additional notions of date semantics).
+5.  ``edx-when``, and the ``schedules`` app, should keep history, so that we can
+    determine at analysis time what their schedule was when they answered a problem.
+    Using a simple history table should be sufficient to also support showing a user
+    the history of changes to their own schedule.
+
+~~~~~~
+Status
+~~~~~~
+
+Proposed
+
+~~~~~~~~~~~~
+Consequences
+~~~~~~~~~~~~
+
+Once implemented, self-paced courses in the LMS would begin to have
+due dates for course content that are consistent between users, but that
+are personalized based on when a user started a course. We would be open
+to exploring future functionality for allowing users to adjust their schedule
+by resetting their effective course start time, or speeding up or slowing
```

Comment:
> adjust their schedule by resetting their effective course start time

Would we have more flexibility long term if instead of marking milestones as "time after course start" and we marked spans of course content between milestones with expected durations? I could imagine wanting to adjust dates based on "now", rate of course completion, and course content "duration" instead of calculating an effective course start time.

Reply:
We chatted about that in the meeting. My concern with setting up multiple sequential blocks of durations is that to find out the due date of content towards the end of the course, you have to get all of the content at the beginning of the course and then add up their durations. That makes `edx-when` significantly harder to write/reason about when talking about efficiently querying across multiple courses. It's possible that there's some fancy (or maybe not so fancy?) SQL magic we could do to get it to do the summing for us, but we'd still need add more information about the course structure/sequence to the data model as well.

In the end, just using `start`, and then back-computing all existing deadlines seemed simpler.

Another factor was the idea that if you have to adjust your schedule (say, slowing down the pacing), then you probably were behind, and thus actually want to go back and adjust old due dates as well, so that you get more time on the content that you missed. If you only adjust future content, then the user has to catch that they are falling behind before they actually miss content that is required.

Reply:
I see. The reasoning makes sense to me, but to just illustrate my thought a little further and I'll let it rest:

My understanding of the proposal: `Due Date = Start Date + Milestone Date`

If a user has completed half the course and falls behind, how do you calculate the new start date such that the new due dates a reasonable? Would you have the user choose a date? Would you calculate a new start date such that the current learning sequence's due date is some reasonable time in the future (may be difficult to write/reason)?

Sequence durations: `Due Date = Now + sum(prior uncompleted sequences)`

If the user has completed half the course and falls behind, the calculations for due dates remains now + sum of prior uncompleted sequences. Everything after and including the current sequence would be recalculated. This wouldn't require user input or any unusual calculations.